state-reconstruct-fetcher/src/blob_http_client.rs

```diff
@@ -0,0 +1,107 @@
+use serde_json::Value;
+use tokio::time::{sleep, Duration};
+
+use crate::types::ParseError;
+
+/// `MAX_RETRIES` is the maximum number of retries on failed blob retrieval.
+const MAX_RETRIES: u8 = 5;
+/// The interval in seconds to wait before retrying to fetch a blob.
+const FAILED_FETCH_RETRY_INTERVAL_S: u64 = 10;
+
+pub struct BlobHttpClient {
+    client: reqwest::Client,
+    url_base: String,
+}
+
+impl BlobHttpClient {
+    pub fn new(blob_url: String) -> eyre::Result<Self> {
+        let mut headers = reqwest::header::HeaderMap::new();
+        headers.insert(
+            "Accept",
+            reqwest::header::HeaderValue::from_static("application/json"),
+        );
+        let client = reqwest::Client::builder()
+            .default_headers(headers)
+            .build()?;
+        Ok(Self {
+            client,
+            url_base: blob_url,
+        })
+    }
+
+    pub async fn get_blob(&self, kzg_commitment: &[u8]) -> Result<Vec<u8>, ParseError> {
+        let url = self.format_url(kzg_commitment);
+        for attempt in 1..=MAX_RETRIES {
+            match self.retrieve_url(&url).await {
+                Ok(response) => match response.text().await {
+                    Ok(text) => match get_blob_data(&text) {
+                        Ok(data) => {
+                            let plain = if let Some(p) = data.strip_prefix("0x") {
+                                p
+                            } else {
+                                &data
+                            };
+                            return hex::decode(plain).map_err(|e| {
+                                ParseError::BlobFormatError(plain.to_string(), e.to_string())
+                            });
+                        }
+                        Err(e) => {
+                            tracing::error!("failed parsing response of {url}");
+                            return Err(e);
+                        }
+                    },
+                    Err(e) => {
+                        tracing::error!("attempt {}: {} failed: {:?}", attempt, url, e);
+                        sleep(Duration::from_secs(FAILED_FETCH_RETRY_INTERVAL_S)).await;
+                    }
+                },
```

Comment:
```suggestion
            Ok(response) => {
                let Ok(text) = response.text().await else {
                    continue;
                };

                let data = get_blob_data(&text)?;
                let plain = if let Some(p) = data.strip_prefix("0x") {
                    p
                } else {
                    &data
                };
                return hex::decode(plain)
                    .map_err(|e| ParseError::BlobFormatError(plain.to_string(), e.to_string())));
            }
```
How does this look? It accomplishes pretty much the same thing as it did previously, but in my opinion it's a lot more readable this way.

Reply:
But it isn't the same thing at all... In case response cannot be retrieved, I want to see the error (I don't think I ever did - I suppose it's there for some complicated HTTP stuff that doesn't really happen, but who knows), not just try again...

Reply:
It will give you the error if the response fails though, that part hasn't been changed at all.

However, another thing I thought of: we should actually be getting the response as a [json](https://docs.rs/reqwest/latest/reqwest/#json) directly instead of getting it as a string and then parsing it as one.

Reply:
> It will give you the error if the response fails though, that part hasn't been changed at all.

Well, no.  Let's say I replace the `Ok(text) = response.text()` with `Err(x) = response.text()`, to simulate the failure (full executable code is in https://github.com/eqlabs/zksync-state-reconstruct/tree/inversion ). Then I can run `cargo run reconstruct l1 --http-url https://eth.llamarpc.com --start-block 19428415` to check some blobs, and get
```
2024-03-21T06:57:31.341353Z  INFO url = https://api.blobscan.com/blobs/0xb1647613a7b52bae4aa8f359421d1c9e8144d62768ac00946afb81838d24a3f76172fdfd01ab09f5bfa2e1638954e513
2024-03-21T06:57:32.394116Z  INFO attempt 1: got response
2024-03-21T06:57:33.340344Z  INFO attempt 2: got response
2024-03-21T06:57:34.429335Z  INFO attempt 3: got response
2024-03-21T06:57:35.041117Z  INFO attempt 4: got response
2024-03-21T06:57:35.635881Z  INFO attempt 5: got response
2024-03-21T06:57:40.777653Z  INFO PROGRESS: [ 0%] CUR BLOCK L1: 19428483 L2: 460962 TOTAL BLOCKS PROCESSED L1: 68 L2: 0
```
No error is logged, and really I don't see how it could...

Reply:
> However, another thing I thought of: we should actually be getting the response as a [json](https://docs.rs/reqwest/latest/reqwest/#json) directly instead of getting it as a string and then parsing it as one.
That link is for sending JSON, but yes, something like https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.json should be possible - except I don't see how to get the response text after failing to parse it as JSON... Really, all I want here is error handling - get response, report error on failure, then parse it, report error on failure... If that's more elaborate than some online Rust examples, well, so what?

Reply:
That code is testing that we can _decode_ the response, not _get_ the response. Getting the response is still handled with a `match` statement like it was previously.

I'll admit, I don't really see why we would need that level of error logging. Failing to parse it as JSON is in my mind more than good enough, since that tells the user exactly what went wrong - it's not a valid JSON response.

And for the error handling part, if you're insistent, I would highly recommend looking into more advanced `thiserror` uses, like the `#[from]` derive. That allows you to do what you want, _without_ polluting the code with so much boilerplate. If you go down this road, I'd also suggest making `BlobError` it's own error type that can then be condensed to a `ParseError`.

Reply:
There are 2 async calls needed to get the response: `send` (inside `BlobHttpClient::retrieve_url`) and `text`. Both can fail, and both should be tested. I want to log the second one, precisely because I do not expect the error to happen.

> I'll admit, I don't really see why we would need that level of error logging. 

Well, to debug the problems _that we've already had_ (e.g. the wrong URL returning JSON without a `data` attribute - looking at it, it's obvious it says NOT_FOUND, but a serde parser will loose that info).

Reply:
Gotcha, sure. in that case we should definitely be using the `?` operator in combination with `thiserror`-style error handling then since that would be an unrecoverable error. See my comment above for details.